src/compiler/crystal/semantic/recursive_struct_checker.cr

```diff
@@ -140,6 +140,6 @@ class Crystal::RecursiveStructChecker
   end
 
   def struct?(type)
-    type.struct? && type.is_a?(InstanceVarContainer) && !type.is_a?(PrimitiveType) && !type.is_a?(ProcInstanceType) && !type.is_a?(GenericClassType) && !type.abstract?
```

Comment:
@asterite the GenericClassType are used to represent generic structs also? Because I don't find right away why there was a `!type.is_a?(GenericClassType)` otherwise.

Reply:
> the GenericClassType are used to represent generic structs also?

Yes

> Because I don't find right away why there was a !type.is_a?(GenericClassType) otherwise.

I can't remember why that condition exists either. But if tests pass then I guess it's OK.